src/pairinteraction/_wrapped/state/state.py

```diff
@@ -0,0 +1,89 @@
+    from pairinteraction._wrapped.ket.ket import KetBase
+    from pairinteraction.units import NDArray
+
+KetType = TypeVar("KetType", bound="KetBase", covariant=True)
+BasisType = TypeVar("BasisType", bound="BasisBase[Any, Any]", covariant=True)
+
+
+class StateBase(ABC, Generic[BasisType, KetType]):
+    """Base class for all State objects.
+
+    The state objects are meant to represent a set of kets, that span a Hilbert space
+    and store a coefficient vector, that describes the state in terms of the kets.
+
+    Basically just a wrapper around the python Basis classes, but with more convenience functions specific to a state.
+
+    Note, in the cpp code we dont have an explicit state class, but use a Basis object, which has only one state
+
+    All state objects share a few common attributes and methods, that are defined in this base class, e.g.:
+        - the number of kets,
+        - the kets of the state,
+        - the coefficient vector as 1d-array,
+        - ...
+    """
+
+    _basis: BasisType
+    _TypeKet: type[KetType]  # should be ClassVar, but cannot be nested yet
+
+    def __init__(self) -> None:
+        raise NotImplementedError(
+            "State objects cannot be created directly, you can use `basis.get_corresponding_state(ket)` instead."
+        )
+
+    @classmethod
+    def _from_basis_object(cls: "type[Self]", basis_obj: "BasisBase[KetType, StateBase[Any, Any]]") -> "Self":
+        obj = cls.__new__(cls)
+        obj._basis = basis_obj  # type: ignore [assignment]
+        return obj
+
+    def __repr__(self) -> str:
+        return self.get_label()
+
+    def __str__(self) -> str:
+        return self.__repr__().replace("Real", "").replace("Complex", "")
+
+    def get_label(self) -> str:
+        """Label representing the state."""
+        coefficients = self.get_coefficients()
+        max_inds = np.argsort(np.abs(coefficients))[::-1]
+        txt = ""
+        overlap = 0
+        for i, ind in enumerate(max_inds):
+            txt += f"{coefficients[ind]:.1f} {self.kets[ind].get_label('ket')}"
+            overlap += abs(coefficients[ind]) ** 2
+            if overlap > 0.9 or i == 2:
+                break
+            txt += " + "
+        return f"{type(self).__name__}({txt})"
+
+    @property
+    def kets(self) -> list[KetType]:
+        """Return a list containing the kets of the basis."""
+        return self._basis.kets
+
+    @property
+    def number_of_kets(self) -> int:
+        return self._basis.number_of_kets
+
+    def get_coefficients(self) -> "NDArray":
+        """Return the coefficients of the state as 1d-array."""
+        return self._basis.get_coefficients().toarray().flatten()
+
+    def get_corresponding_ket(self) -> KetType:
+        cpp_ket = self._basis._cpp.get_corresponding_ket(self._basis._cpp)  # type: ignore [arg-type]
+        return self._TypeKet._from_cpp_object(cpp_ket)
+
+    def get_corresponding_ket_index(self) -> int:
```

Comment:
Hmm, here are corresponding_ket methods but not in the Basis class. Probably, I overlooked something.